Add tracing to image_pipeline and instrument two ROS 2 components

vmayoral opened the pull request:
* Fork [tracetools](https://gitlab.com/ros-tracing/ros2_tracing/-/tree/master/tracetools) and shape it as `tracetools_image_pipeline` to avoid external dependencies.
* Tracing first examples: `image_proc::RectifyNode` and `image_proc::ResizeNode`

vmayoral committed:
Add tracetools_image_pipeline package to the pipeline

Add an LTTng tracing provider wrapper for image_pipeline
metapackage. Include tracepoints for two initial ROS 2
components: image_proc::RectifyNode and ResizeNode

Signed-off-by: Víctor Mayoral Vilches <v.mayoralv@gmail.com>

vmayoral committed:
LTTng instrument image_proc::RectifyNode and image_proc::ResizeNode

Signed-off-by: Víctor Mayoral Vilches <v.mayoralv@gmail.com>

vmayoral commented:
Ping @christophebedard, I'd like you to confirm you're fine with this.

Note that I've forked [tracetools](https://gitlab.com/ros-tracing/ros2_tracing/-/tree/master/tracetools) and instead, this PR proposes to create a new ROS 2 package `tracetools_image_pipeline` which includes **only** the `TRACEPOINTS` for this meta-package and which can be maintained more easily (instead of having to send PRs to [tracetools](https://gitlab.com/ros-tracing/ros2_tracing/-/tree/master/tracetools) and iterate in longer cycles).

The rationale behind this comes from the the experience of having worked for a few months with the ROS 2 tracing infrastructure. To me, it eventually become rather cumbersome to have to *maintain* a single *tracetools* repo with all tracepoints (I experienced this directly with an extension of [tracetools_acceleration](https://github.com/ros-acceleration/tracetools_acceleration) we keep internally). Altogether this led me to realize that it might be *better* to simply add a `tracetools_<whatever-ros2-stack/package>`.

@vrabaud and @JWhitleyWork, note that this contribution adds the new package proposed [as a `REQUIRED` dependency](https://github.com/ros-perception/image_pipeline/pull/717/files#diff-9c994a514e4363196b68034262271d054182fc9e0682f9f11e2d2646970aa9c6R16), if LTTng is not installed into the system, it should nicely inform that tracing is disabled but packages should build still fine. For now only `image_proc::RectifyNode` and `image_proc::ResizeNode` have been instrumented but if this approach receives a 👍, happy to instrument a few more components of the pipeline.

A simple launch file to test it:

```python
"""Launch a raw image publisher, pipelined with rectify and resize"""

import os
import rclpy
import launch
from rclpy.node import Node
from std_msgs.msg import String
from sensor_msgs.msg import CameraInfo, Image
from launch_ros.actions import ComposableNodeContainer, Node
from launch_ros.descriptions import ComposableNode
from tracetools_launch.action import Trace
from tracetools_trace.tools.names import DEFAULT_EVENTS_ROS


def generate_launch_description():
    """Generate launch description with multiple components."""

    # # An image_raw publisher
    # image_raw_publisher = Node(
    #     package="perception_2nodes",
    #     executable="image_raw_publisher",
    #     name="image_raw_publisher",
    #     remappings=[("image_raw", "image")],
    # )

    # Trace
    trace = Trace(
        session_name="raw_rectify_resize_pipeline",
        events_ust=[
            "ros2_image_pipeline:*",
        ]
        + DEFAULT_EVENTS_ROS,
    )

    perception_container = ComposableNodeContainer(
        name="perception_container",
        namespace="",
        package="rclcpp_components",
        executable="component_container",
        composable_node_descriptions=[
            ComposableNode(
                package="image_proc",
                plugin="image_proc::RectifyNode",
                name="rectify_node",
            ),
            ComposableNode(
                namespace="resize",
                package="image_proc",
                plugin="image_proc::ResizeNode",
                name="resize_node",
                remappings=[
                    ("camera_info", "/camera_info"),
                    ("image", "/image_rect"),
                    ("resize", "resize"),
                ],
                parameters=[
                    {
                        "scale_height": 2.0,
                        "scale_width": 2.0,
                    }
                ],
            ),
        ],
        output="screen",
    )
    # return launch.LaunchDescription([image_raw_publisher, perception_container, trace])
    return launch.LaunchDescription([perception_container, trace])

```

vmayoral committed:
Deal with uncrustify and cpplint

Signed-off-by: Víctor Mayoral Vilches <v.mayoralv@gmail.com>

christophebedard commented:
> Ping @christophebedard, I'd like you to confirm you're fine with this.
> 
> Note that I've forked [tracetools](https://gitlab.com/ros-tracing/ros2_tracing/-/tree/master/tracetools) and instead, this PR proposes to create a new ROS 2 package `tracetools_image_pipeline` which includes **only** the `TRACEPOINTS` for this meta-package and which can be maintained more easily (instead of having to send PRs to [tracetools](https://gitlab.com/ros-tracing/ros2_tracing/-/tree/master/tracetools) and iterate in longer cycles).
> 
> The rationale behind this comes from the the experience of having worked for a few months with the ROS 2 tracing infrastructure. To me, it eventually become rather cumbersome to have to _maintain_ a single _tracetools_ repo with all tracepoints (I experienced this directly with an extension of [tracetools_acceleration](https://github.com/ros-acceleration/tracetools_acceleration) we keep internally). Altogether this led me to realize that it might be _better_ to simply add a `tracetools_<whatever-ros2-stack/package>`.

Happy to see people extending/building on top of this @vmayoral! :smiley:

This looks good to me. For reference, a similar rationale wrt the ROS 2 core itself was brought up a few months ago here: https://github.com/ros2/rclcpp/pull/1738#issuecomment-892934136. If these were tracepoints for the ROS 2 *core* (i.e., currently `rclcpp*` and below), I'd prefer to avoid forking `tracetools`/putting the tracepoints in other packages to avoid the usual fragmentation of open source code, but this isn't really the case. It's more of an extension, i.e., additional instrumentation for application-level layers above the core, so I think this is the way to go for these use-cases. Besides, there is nothing about LTTng that requires instrumentation (or, rather, tracepoints) to be centralized like it is currently for the ROS 2 core, and the interactions with the `Trace` action are not really affected by this choice.

I'm not entirely sure whether or not these "instrumentation extension" packages should use the same `TRACETOOLS_*` CMake options, but I think it does make sense to use the same option name for all of them and have globals options. Besides, a package-specific option can always be added later on.

vmayoral commented:
Happy new year folks! 

I'm bumping this ticket so that it doesn't get stalled. @vrabaud and @JWhitleyWork, you guys are listed as maintainers, @jacobperron you seem to have contributed recently as well, anything else needed from our side on this PR? With it, we should be able to add additional probes.

jacobperron commented on tracetools_image_pipeline/CHANGELOG.rst:
```diff
@@ -0,0 +1,10 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Changelog for package tracetools_image_pipeline
```
I think we can omit the changelog file. This is typically generated at the time of a release.

vmayoral replied:
Thanks @jacobperron, see 6c3a016757853316c2293bc84a978d2ae120e7bb

jacobperron commented on tracetools_image_pipeline/package.xml:
```diff
@@ -0,0 +1,23 @@
+<?xml-model href="http://download.ros.org/schema/package_format2.xsd" schematypens="http://www.w3.org/2001/XMLSchema"?>
+<package format="2">
+  <name>tracetools_image_pipeline</name>
+  <version>0.1.0</version>
```
This version should be `2.2.1`, matching other packages in this repo.

vmayoral replied:
Sure ee087e7cfddea1718a6c61a410db3a78dfdfbb9a.

jacobperron reviewed (commented):
Looks okay to me. I didn't test it, but I think @JWhitleyWork should give the final approval.

JWhitleyWork commented on image_proc/CMakeLists.txt:
```diff
@@ -13,6 +13,7 @@ endif()
 find_package(ament_cmake_auto REQUIRED)
 ament_auto_find_build_dependencies()
 
+find_package(tracetools_image_pipeline REQUIRED)
```
`ament_auto_find_build_dependencies()` will automatically add this if it is in `package.xml` so this line can be removed.

vmayoral replied:
Right! Thanks 621d0ffabec44e8f3b2db1812f8787c8acfbeabe.

JWhitleyWork commented:
Please address the 3 changes above (mine and @jacobperron's) and then I'm good to merge.

vmayoral committed:
Omit changelogfile

See https://github.com/ros-perception/image_pipeline/pull/717#discussion_r781637389

Signed-off-by: Víctor Mayoral Vilches <v.mayoralv@gmail.com>

vmayoral committed:
tracetools_image_pipeline version consistent with repo

Signed-off-by: Víctor Mayoral Vilches <v.mayoralv@gmail.com>

vmayoral committed:
Remove unnecessary find_package

tracetools_image_pipeline included in package.xml and
fetched by ament_auto_find_build_dependencies().

Signed-off-by: Víctor Mayoral Vilches <v.mayoralv@gmail.com>

vmayoral commented:
All done from my side @jacobperron and @JWhitleyWork, back to you.

vmayoral commented:
Friendly ping @JWhitleyWork
